### PR TITLE
OOPs Tracer Remapping

### PR DESCRIPTION
**Pull request by oelbert**
## Purpose

This PR oopifies mapn_tracer

## Code changes:

- mapn_tracer.py: created MapNTracer class to cache stencils and variables
- map_single.py: expanded docstrings
- remapping_part1.py: changed the call to mapn_tracer to use the new class
- translate_mapn_tracer_2d.py: changed the call to mapn_tracer to use the new class

## Checklist
Before submitting this PR, please make sure:

- [x] You have followed the coding standards guidelines established at [Code Review Checklist](https://paper.dropbox.com/doc/Code-Review-Checklist--BD7zigBMAhMZAPkeNENeuU2UAg-IlsYffZgTwyKEylty7NhY).
- [x] Docstrings and type hints are added to new and updated routines, as appropriate
- [x] All relevant documentation has been updated or added (e.g. README, CONTRIBUTING docs)
- [x] Unit tests are added or updated for non-stencil code changes

### CODE REVIEW

File: fv3core/stencils/map_single.py [hunk 1]
```diff
@@ -126,6 +126,19 @@ def __call__(
         q4_4: FloatField,
         dp1: FloatField,
     ):
+        """
```
Comment (eddie-c-davis) [inline fv3core/stencils/map_single.py:129]: Thank you for updating the doc string. Perhaps you can show me where you find this information :)

File: fv3core/stencils/mapn_tracer.py [hunk 2]
```diff
@@ -7,31 +7,68 @@
...
+        self._i2 = i2
+        self._j1 = j1
+        self._j2 = j2
+        self._qs = utils.make_storage_from_shape(
+            (grid.npx, grid.npy, self._nk + 1),
+            origin=(0, 0, 0),
+            cache_key="mapn_tracer_qs",
+        )
+
+        self._map_single = utils.cached_stencil_class(MapSingle)(
```
Comment (eddie-c-davis) [inline fv3core/stencils/mapn_tracer.py:30]: Do we still need cache keys for storages and stencils that are object members?

File: fv3core/stencils/mapn_tracer.py [hunk 3]
```diff
@@ -7,31 +7,68 @@
...
+            kord, 0, i1, i2, j1, j2, cache_key=f"mapntracer-single{j2}"
         )
-        fillz(dp2, tracers, map_single.i_extent, map_single.j_extent, spec.grid.npz, nq)
+
+        if spec.namelist.fill:
+            self._fill_negative_tracers = True
+            self._fillz = utils.cached_stencil_class(FillNegativeTracerValues)(
```
Comment (eddie-c-davis) [inline fv3core/stencils/mapn_tracer.py:36]: Same as above.

File: fv3core/stencils/mapn_tracer.py [hunk 4]
```diff
@@ -7,31 +7,68 @@
...
+        self,
+        pe1: FloatField,
+        pe2: FloatField,
+        dp2: FloatField,
+        tracers: Dict[str, "FloatField"],
+        nq: int,
+        q_min: float,
+    ):
+        """
+        Remaps the tracer species onto the Eulerian grid
```
Comment (eddie-c-davis) [inline fv3core/stencils/mapn_tracer.py:52]: Again, thank you for the doc string!

File: fv3core/stencils/mapn_tracer.py [hunk 5]
```diff
@@ -7,31 +7,68 @@
...
+            pe2 (out): Eulerian pressure levels
+            dp2 (in)
+            qs (out): Field to be remapped on deformed grid
+            jfirst: Starting index of the J-dir compute domain
+            jlast: Final index of the J-dir compute domain
+        """
+        for q in utils.tracer_variables[0:nq]:
+            self._map_single(tracers[q], pe1, pe2, self._qs)
+
+        if self._fill_negative_tracers is True:
```
Comment (eddie-c-davis) [inline fv3core/stencils/mapn_tracer.py:66]: Would `if self._fill_negative_tracers:` suffice?

File: tests/translate/translate_mapn_tracer_2d.py [hunk 6]
```diff
@@ -42,5 +41,12 @@ def compute(self, inputs):
             pad_field_in_j(inputs["dp2"], self.grid.npy)
         )
         inputs["kord"] = abs(spec.namelist.kord_tr)
+        self.compute_func = MapN_Tracer.MapNTracer(
```
Comment (eddie-c-davis) [inline tests/translate/translate_mapn_tracer_2d.py:44]: Nice and clean...

File: fv3core/stencils/remapping_part1.py [hunk 7]
```diff
@@ -303,19 +303,15 @@ def compute(
...
 
     # TODO if nq > 5:
-    mapn_tracer.compute(
-        pe1,
-        pe2,
-        dp2,
-        tracers,
-        nq,
-        0.0,
+    MapN_Tracer = utils.cached_stencil_class(MapNTracer)(
```
Comment (eddie-c-davis) [inline fv3core/stencils/remapping_part1.py:306]: Should this begin with a lowercase `m`, since it is a variable name?

**Review by eddie-c-davis — CHANGES_REQUESTED** [review]
A few small (admittedly somewhat nitpicky) suggestions...

File: fv3core/stencils/mapn_tracer.py (hunk 2 above)
Comment (oelbert) [reply fv3core/stencils/mapn_tracer.py:30]: Nope, I just neglected to update that

File: fv3core/stencils/map_single.py (hunk 1 above)
Comment (oelbert) [reply fv3core/stencils/map_single.py:129]: Mostly just from porting the Fortran code. Hopefully the documentation will come out soon and provide a good reference for this

File: fv3core/stencils/mapn_tracer.py (hunk 5 above)
Comment (oelbert) [reply fv3core/stencils/mapn_tracer.py:66]: I've had some conversations about avoiding that pattern elsewhere in the code, so I was just trying to keep it consistent here.

File: fv3core/stencils/remapping_part1.py (hunk 7 above)
Comment (oelbert) [reply fv3core/stencils/remapping_part1.py:306]: yes, thanks!

File: tests/translate/translate_mapn_tracer_2d.py (hunk 6 above)
Comment (oelbert) [reply tests/translate/translate_mapn_tracer_2d.py:44]: This feels very Pythonic to me :-)

**Comment by eddie-c-davis** [comment]
launch jenkins

**Review by eddie-c-davis — APPROVED** [review]
Looks great, well done!

**Comment by oelbert** [comment]
launch jenkins

**Pull request merged**